views.py

```diff
@@ -35,12 +40,38 @@ def get_context_data(self, **kwargs):
             'portal_id': self.hubspot_portal_id,
             'form_id': self.hubspot_form_id,
         }
+        kwargs['language_options'] = self.language_options()
         return super(BasePreloginView, self).get_context_data(**kwargs)
 
     @use_bootstrap3
     def dispatch(self, request, *args, **kwargs):
         return super(BasePreloginView, self).dispatch(request, *args, **kwargs)
 
+    def get(self, request, *args, **kwargs):
+        lang_prefix = kwargs.get('lang_code')
+        if lang_prefix and lang_prefix not in prelogin_lang_codes():
+            return HttpResponseRedirect(reverse(self.urlname))
+        self.activate_language_from_request(request, lang_prefix)
+        return super(BasePreloginView, self).get(request, *args, **kwargs)
+
+    def activate_language_from_request(self, request, lang_prefix):
+        if lang_prefix in prelogin_lang_codes():
+            lang = lang_prefix
+        else:
+            lang = translation.get_language_from_request(request)
+        request.session[translation.LANGUAGE_SESSION_KEY] = lang
+        request.session['django_language'] = lang
+        translation.activate(lang)
+
+    def language_options(self):
+        options = []
+        for lang_code, display_label in PRELOGIN_LANGUAGES:
+            options.append({
+                'display_label': display_label,
+                'prefix_url': reverse(self.urlname, args=[lang_code])
+            })
+        return options
```

Comment:
not blocking but this whole function looks like it could be replaced with a list comprehension which might be more readable